src/query/storage/prom_converter.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/cost"
+	"github.com/m3db/m3/src/query/generated/proto/prompb"
+	"github.com/m3db/m3/src/query/models"
+	xcost "github.com/m3db/m3/src/x/cost"
+	"github.com/m3db/m3/src/x/ident"
+	xsync "github.com/m3db/m3/src/x/sync"
+)
+
+func tagIteratorToLabels(
+	identTags ident.TagIterator,
+) ([]prompb.Label, error) {
+	labels := make([]prompb.Label, 0, identTags.Remaining())
+	for identTags.Next() {
+		identTag := identTags.Current()
+		labels = append(labels, prompb.Label{
+			Name:  identTag.Name.Bytes(),
+			Value: identTag.Value.Bytes(),
+		})
+	}
+
+	if err := identTags.Err(); err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+func iteratorToPromResult(
+	iter encoding.SeriesIterator,
+	enforcer cost.ChainedEnforcer,
+	tagOptions models.TagOptions,
+) (*prompb.TimeSeries, error) {
+	labels, err := tagIteratorToLabels(iter.Tags())
+	if err != nil {
+		return nil, err
+	}
+
+	samples := make([]prompb.Sample, 0, initRawFetchAllocSize)
+	for iter.Next() {
+		dp, _, _ := iter.Current()
+		samples = append(samples, prompb.Sample{
+			Timestamp: TimeToPromTimestamp(dp.Timestamp),
+			Value:     dp.Value,
+		})
+	}
+
+	if err := iter.Err(); err != nil {
+		return nil, err
+	}
+
+	r := enforcer.Add(xcost.Cost(len(samples)))
+	if r.Error != nil {
+		return nil, r.Error
+	}
+
+	return &prompb.TimeSeries{
+		Labels:  labels,
+		Samples: samples,
+	}, nil
+}
+
+// Fall back to sequential decompression if unable to decompress concurrently.
+func toPromSequentially(
+	iters []encoding.SeriesIterator,
+	enforcer cost.ChainedEnforcer,
+	metadata block.ResultMetadata,
+	tagOptions models.TagOptions,
+) (PromResult, error) {
+	seriesList := make([]*prompb.TimeSeries, 0, len(iters))
+	for _, iter := range iters {
+		series, err := iteratorToPromResult(iter, enforcer, tagOptions)
+		if err != nil {
+			return PromResult{}, err
+		}
+
+		if len(series.GetSamples()) == 0 {
+			continue
+		}
+
+		seriesList = append(seriesList, series)
+	}
+
+	return PromResult{
+		Metadata: metadata,
+		PromResult: &prompb.QueryResult{
+			Timeseries: seriesList,
+		},
+	}, nil
+}
+
+func toPromConcurrently(
+	iters []encoding.SeriesIterator,
+	readWorkerPool xsync.PooledWorkerPool,
+	enforcer cost.ChainedEnforcer,
+	metadata block.ResultMetadata,
+	tagOptions models.TagOptions,
+) (PromResult, error) {
+	seriesList := make([]*prompb.TimeSeries, len(iters))
+	errorCh := make(chan error, 1)
+	done := make(chan struct{})
+	stopped := func() bool {
+		select {
+		case <-done:
+			return true
+		default:
+			return false
+		}
+	}
+
+	var wg sync.WaitGroup
+	for i, iter := range iters {
+		i, iter := i, iter
+		wg.Add(1)
+		readWorkerPool.Go(func() {
+			defer wg.Done()
+			if stopped() {
+				return
+			}
+
+			series, err := iteratorToPromResult(iter, enforcer, tagOptions)
+			if err != nil {
+				// Return the first error that is encountered.
+				select {
+				case errorCh <- err:
```

Comment:
Hm, seems racey here that errorCh might be already closed? 

This is since stopped() is not read or any lock is held to guarantee the errorCh is still open before enqueuing. 

Unfortunately even in a switch like this writing to a closed channel will panic:
https://play.golang.org/p/XmRlQbixatB

Can use a cancellable lifetime here if you like:
https://github.com/m3db/m3/blob/master/src/x/resource/lifetime.go#L28:6